### PR TITLE
[ASDisplayNode] Add support for setting non-fatal error block

### PR DESCRIPTION
**Pull request by rmls**
Currently Async Display Kit doesn't provide a mechanism for "asserting" in production. In many the framework can recover from failures, but that would happen silently when in production.

This code changes gives the ability to set a "non-fatal error block" on `ASDisplayNode` class, so it can be plugged in many areas where we have asserts, for example, so developers have the change to capture and report such failures in production.

This addresses #2807 .

### CODE REVIEW

File: AsyncDisplayKit/ASDisplayNode.h [hunk 1]
```diff
@@ -53,6 +53,12 @@ typedef void (^ASDisplayNodeContextModifier)(_Nonnull CGContextRef context);
 typedef ASLayoutSpec * _Nonnull(^ASLayoutSpecBlock)(__kindof ASDisplayNode * _Nonnull node, ASSizeRange constrainedSize);
 
 /**
+ * AsyncDisplayKit non-faltal error block. This block can be used for handling non-fatal errors. Useful for reporting
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode.h]: Typo: "non-faltal"

File: AsyncDisplayKit/ASDisplayNode.h [hunk 2]
```diff
@@ -256,6 +262,15 @@ NS_ASSUME_NONNULL_BEGIN
  */
 @property (readonly) ASInterfaceState interfaceState;
 
+/**
+ * @abstract Class property that allows to set a block that can be called on non-fatal errors. This
+ * property can be useful for cases when Async Display Kit can recover from an anormal behavior, but
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode.h]: typo: "anormal"

File: AsyncDisplayKit/ASDisplayNode.h [hunk 3]
```diff
@@ -256,6 +262,15 @@ NS_ASSUME_NONNULL_BEGIN
...
+ * still gives the opportunity to use a reporting mechanism to catch occurrences in production.
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode.h]: Maybe add: "In development, ASDK will assert instead of calling this."

**Comment by Adlai-Holler** [comment]
Jenkins, add to whitelist

File: Base/ASAssert.h [hunk 4]
```diff
@@ -59,3 +59,17 @@
 #define ASDisplayNodeCAssertPositiveReal(description, num) ASDisplayNodeCAssert(num >= 0 && num <= CGFLOAT_MAX, @"%@ must be a real positive integer.", description)
 #define ASDisplayNodeCAssertInfOrPositiveReal(description, num) ASDisplayNodeCAssert(isinf(num) || (num >= 0 && num <= CGFLOAT_MAX), @"%@ must be infinite or a real positive integer.", description)
 
+#define ASDisplayNodeErrorDomain @"ASDisplayNodeErrorDomain"
+#define ASDisplayNodeNonFatalErrorCode 0
```
Comment (garrettmoon) [inline Base/ASAssert.h]: Total nit, but maybe make this `1`? Many systems use zero as a non-error.

File: Base/ASAssert.h [hunk 5]
```diff
@@ -59,3 +59,17 @@
...
+
+#define ASDisplayNodeAssertNonFatal(desc, ...)                                                                                  \
+  ASDisplayNodeAssert(NO, desc, ##__VA_ARGS__);                                                                                 \
```
Comment (garrettmoon) [inline Base/ASAssert.h]: I was thinking more along the lines of this above
```
#define ASDisplayNodeAssertNonFatal(condition, desc, ...)
ASDisplayNodeAssert(condition, desc, ##__VA_ARGS__);
if (condition) {
```

**Pull request merged**